Std/Data/Array/Lemmas.lean

```diff
@@ -336,3 +337,31 @@ theorem forIn_eq_data_forIn [Monad m]
       simp [← this]; congr; funext x; congr; funext b
       rw [loop (i := i)]; rw [← ij, Nat.succ_add]; rfl
   simp [forIn, Array.forIn]; rw [loop (Nat.zero_add _)]; rfl
+
+/-! ### join -/
+
+@[simp] theorem join_data {l : Array (Array α)} : l.join.data = (l.data.map data).join := by
+  dsimp [join]
+  simp only [foldl_eq_foldl_data]
+  generalize l.data = l
+  have : ∀ a : Array α, (List.foldl ?_ a l).data = a.data ++ ?_ := ?_
+  exact this #[]
+  induction l with
+  | nil => simp_all
+  | cons h t ih =>
+    simp only [List.foldl_cons, ih, List.map_cons, List.join_cons, ← List.append_assoc]
+    induction h.data <;> simp_all
```

Comment:
```suggestion
  | nil => simp
  | cons h => induction h.data <;> simp [*]
```
Might depend on new `join` def.

Reply:
Haha, worked both before and after, unlike my proof.